### PR TITLE
[9.0][ADD] Add report_xlsx module

### PR DESCRIPTION
**Pull request by adrienpeiffer**

### CODE REVIEW

File: base_report_xlsx/report/report_xlsx.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
+# -*- coding: utf-8 -*-
+# Copyright 2015 ACSONE SA/NV (<http://acsone.eu>)
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl.html).
+
+import xlsxwriter
```
Comment (pedrobaeza) [inline base_report_xlsx/report/report_xlsx.py]: Please protect import

File: base_report_xlsx/README.rst [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+This module was written to provide a basic report class to generate xlsx report.
+
+Usage
+=====
+
+.. image:: https://odoo-community.org/website/image/ir.attachment/5784_f2813bd/datas
+   :alt: Try me on Runbot
+   :target: https://runbot.odoo-community.org/runbot/143/9.0
+
+For further information, please visit:
```
Comment (pedrobaeza) [inline base_report_xlsx/README.rst]: Don't put this

**Comment by pedrobaeza** [comment]
Can you explain how to use it in the README or put any sample in demo data?

You should also call it _report_xlsx_ as the rest of this kind of modules. An Excel XLSX icon would be also appreciated.

File: base_report_xlsx/README.rst [hunk 3]
```diff
@@ -0,0 +1,49 @@
+.. image:: https://img.shields.io/badge/licence-AGPL--3-blue.svg
+    :target: http://www.gnu.org/licenses/agpl-3.0-standalone.html
+    :alt: License: AGPL-3
+
+================
+Base report xlsx
+================
+
+This module was written to provide a basic report class to generate xlsx report.
```
Comment (elicoidal) [inline base_report_xlsx/README.rst]: This module provides a basic report class to generate xlsx report.

File: base_report_xlsx/README.rst [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+.. image:: https://odoo-community.org/logo.png
+   :alt: Odoo Community Association
+   :target: https://odoo-community.org
+
+This module is maintained by the OCA.
+
+OCA, or the Odoo Community Association, is a nonprofit organization whose mission is to support the collaborative development of Odoo features and promote its widespread use.
+
+To contribute to this module, please visit http://odoo-community.org.
```
Comment (elicoidal) [inline base_report_xlsx/README.rst]: https

**Comment by chermed** [comment]
Unknown report type: xlsx: (<type 'exceptions.NotImplementedError'>, NotImplementedError(u'Unknown report type: xlsx',), <traceback object at 0x7fa50cc8bf38>)

I don't see the inheritance of report type selection `report_type = fields.Selection(selection_add=[('xlsx','Xlsx')])` and Qweb redirection !

**Comment by adrienpeiffer** [comment]
@chermed can you give me the steps to reproduce it ?

**Comment by adrienpeiffer** [comment]
@pedrobaeza Thanks ! I added an exemple in the README file. About the module name, personally I think `base_report_xlsx` it's more suitable considering that it is only a module that provides base definitions for xlsx report.

**Comment by pedrobaeza** [comment]
Please reconsider again the name. Standard modules are called _report_ and _report_webkit_, not _base_report_ and _base_report_webkit_, although they provide base functionality.

**Comment by StefanRijnhart** [comment]
Haven't tested it, but what @chermed says makes sense: report_type is a fields.Selection, so you'd have to add a selection key for 'xlsx' either in this module or in the module that adds an actual report (cf. https://github.com/OCA/reporting-engine/blob/8.0/report_xml/models.py#L15). Makes more sense to do it here.

**Comment by bealdav** [comment]
Thanks a lot for this really valuable contribution. What are the steps for merge, now ?

**Comment by pedrobaeza** [comment]
I have a concern with the chosen name, and the functionality should be tested.

**Comment by adrienpeiffer** [comment]
Hi,
I'll change the name of this module and add 'xlsx' in report type selection. 
Thanks !

**Comment by adrienpeiffer** [comment]
@pedrobaeza @StefanRijnhart It's done. Thanks for the review :smile:

File: report_xlsx/__openerp__.py [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+{
+    'name': "Base report xlsx",
+
+    'summary': """
+        Base module to create xlsx report""",
+    'author': 'ACSONE SA/NV,'
+              'Odoo Community Association (OCA)',
+    'website': "http://acsone.eu",
+    'category': 'Reporting',
+    'version': '8.0.1.0.0',
```
Comment (sbidoul) [inline report_xlsx/__openerp__.py]: 9.0.1.0.0 since you are targeting the 9.0 branch

**Comment by sbidoul** [comment]
A small pylint error: `report_xlsx/models/ir_report.py:5:1: F401 'api' imported but unused`

File: .travis.yml [hunk 6]
```diff
@@ -22,6 +22,7 @@ virtualenv:
 
 install:
   - pip install xlwt
+  - pip install xlsxwriter
```
Comment (sbidoul) [inline .travis.yml]: In the 8.0 branch these are in `requirements.txt` at repo root.

File: report_xlsx/report/report_xlsx.py [hunk 7]
```diff
@@ -24,11 +24,15 @@ def create(self, cr, uid, ids, data, context=None):
...
+                return self.create_xlsx_report(ids, data, report)
         return super(ReportXlsx, self).create(cr, uid, ids, data, context)
 
-    def create_xlsx_report(self, data, objs):
+    def create_xlsx_report(self, ids, data, report):
+        self.parser_instance = self.parser(
+            self.env.cr, self.env.uid, self.name2, self.env.context)
+        objs = self.getObjects(
+            self.env.cr, self.env.uid, ids, self.env.context)
+        self.parser_instance.set_context(objs, data, objs.ids, 'xlsx')
```
Comment (sbidoul) [inline report_xlsx/report/report_xlsx.py]: @sebalix out of curiosity, in which use case does this help?

Also, perhaps `s/obj.ids/ids/` to shave a few milliseconds?

Comment (sebalix) [inline report_xlsx/report/report_xlsx.py]: About the signature, it follows the standard pattern used by Odoo: https://github.com/OCA/OCB/blob/9.0/openerp/report/report_sxw.py#L371

Then, objects are instanciated inside the method thanks to the already defined `getObjects` method.

About the parser class, it is used to inject data or functions to the report. Having the parser instance in the method which generates the XLSX file allows to access these data/functions easily. The `report_xls` module send this parser instance as a parameter: https://github.com/OCA/reporting-engine/blob/8.0/report_xls/report_xls.py#L169 , maybe it's better to do it this way?

Comment (sbidoul) [inline report_xlsx/report/report_xlsx.py]: I understand, thanks!

**Comment by sbidoul** [comment]
:+1:

**Comment by sbidoul** [comment]
Runbot is red because of the missing xslxwriter python dependency

**Comment by sebalix** [comment]
:+1:

**Comment by dreispt** [comment]
AFAICS this is ready to merge.

**Comment by bealdav** [comment]
- 1 should be at the top of the page ;-) thanks

**Comment by dreispt** [comment]
@bealdav Last discussion I followed about that concluded otherwise. Maybe I missed something.

**Comment by bealdav** [comment]
@dreispt , I have not see this discussion, sorry. A link ?

**Comment by dreispt** [comment]
@bealdav [here](https://odoo-community.org/groups/contributors-15/contributors-36863?mode=thread&date_begin=&date_end=)

**Comment by sebastienbeau** [comment]
runbot is green !

**Comment by adrienpeiffer** [comment]
Thanks @sebastienbeau  !

**Pull request merged**